File: app/entity/revision.go
package entity

import (
	"time"

	"github.com/music-gang/music-gang-api/app/apperr"
)

// RevisionVersion indicates the version of revision management system.
type RevisionVersion string

// This is a list of all revisions versions.
const (
	AnchorageVersion RevisionVersion = "Anchorage"

	CurrentRevisionVersion RevisionVersion = AnchorageVersion
)

// RevisionNumber is the revision number of the entity.
type RevisionNumber uint

// Revisions is a slice of Revision.
type Revisions []*Revision

// Revision represents a revision of a contract.
// Each revision is a snapshot of the contract state.
type Revision struct {
	ID           int64           `json:"id"`
	CreatedAt    time.Time       `json:"created_at"`
	Rev          RevisionNumber  `json:"rev"`
	Version      RevisionVersion `json:"version"`
	ContractID   int64           `json:"contract_id"`
	Notes        string          `json:"notes"`
	CompiledCode []byte          `json:"-"`
	MaxFuel      Fuel            `json:"max_fuel"`

	Contract *Contract `json:"contract"`
}

// Validate validates the revision.
func (r *Revision) Validate() error {

	if r.Rev == 0 {

		return apperr.Errorf(apperr.EINVALID, "revision number cannot be zero")

	} else if r.ContractID == 0 {

		return apperr.Errorf(apperr.EINVALID, "contract id is required")

	} else if r.Version != CurrentRevisionVersion {

		return apperr.Errorf(apperr.EINVALID, "invalid revision version")

	} else if r.CreatedAt.IsZero() {

		return apperr.Errorf(apperr.EINVALID, "created at is required")

	} else if r.MaxFuel == 0 {

		return apperr.Errorf(apperr.EINVALID, "max fuel is required")

	} else if len(r.CompiledCode) == 0 {

		return apperr.Errorf(apperr.EINVALID, "compiled code is required")
	}

	return nil
}

File: tests/entity/revision.rb
# frozen_string_literal: true

# Revision entity.
class Revision
  attr_accessor :id, :created_at, :rev, :version, :contract_id, :notes, :max_fuel

  def initialize(id: nil,
                 created_at: nil,
                 rev: nil,
                 version: nil,
                 contract_id: nil,
                 notes: nil,
                 max_fuel: nil)
    @id = id
    @created_at = created_at
    @rev = rev
    @version = version
    @contract_id = contract_id
    @notes = notes
    @max_fuel = max_fuel
  end

  class << self
    # from_hash
    # @param hash [Hash]
    def from_hash(hash)
      validate_hash hash
      Revision.new id: hash[:id], created_at: Time.parse(hash[:created_at]), rev: hash[:rev], version: hash[:version], contract_id: hash[:contract_id], notes: hash[:notes], max_fuel: hash[:max_fuel]
    end

    # Validate hash.
    # @param hash [Hash]
    def validate_hash(hash)
      raise 'missing id' unless hash.key? :id
      raise 'missing created at' unless hash.key? :created_at
      raise 'missing rev' unless hash.key? :rev
      raise 'missing version' unless hash.key? :version
      raise 'missing contract id' unless hash.key? :contract_id
      raise 'missing notes' unless hash.key? :notes
      raise 'missing max fuel' unless hash.key? :max_fuel
    end
  end
end
